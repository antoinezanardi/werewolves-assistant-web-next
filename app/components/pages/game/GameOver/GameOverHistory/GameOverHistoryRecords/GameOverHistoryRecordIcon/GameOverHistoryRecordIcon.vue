<template>
  <div
    id="game-history-record-icon"
    class="align-items-center flex h-12 justify-content-center shadow-1 text-white w-12 z-1"
  >
    <NuxtImg
      :alt="$t(`components.GameOverHistoryRecordIcon.gameHistoryRecordIconAltText`)"
      placeholder="/svg/misc/ripples.svg"
      :src="gameHistoryRecordIconPath"
    />
  </div>
</template>

<script setup lang="ts">
import type { GameOverHistoryRecordProps } from "~/components/pages/game/GameOver/GameOverHistory/GameOverHistoryRecords/GameOverHistoryRecord/game-over-history-record.types";
import { GAME_PLAYS_TEXT_AND_SVG } from "~/components/shared/game/game-play/game-play.constants";

const props = defineProps<GameOverHistoryRecordProps>();

const gameHistoryRecordIconPath = computed<string>(() => {
  const { gameHistoryRecord } = props;
  const { source, action } = gameHistoryRecord.play;

  const svgPath = GAME_PLAYS_TEXT_AND_SVG[source.name][action]?.svgPath;

  return svgPath ?? "/svg/misc/question-mark.svg";
});
</script>