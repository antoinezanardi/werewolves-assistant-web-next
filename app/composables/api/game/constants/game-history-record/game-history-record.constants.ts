const GAME_HISTORY_RECORD_PLAY_VOTING_RESULTS = [
  "sheriff-election",
  "tie",
  "death",
  "inconsequential",
  "skipped",
] as const;

const GAME_HISTORY_RECORD_PLAYER_ATTRIBUTE_ALTERATION_STATUSES = [
  "attached",
  "detached",
  "activated",
] as const;

export {
  GAME_HISTORY_RECORD_PLAY_VOTING_RESULTS,
  GAME_HISTORY_RECORD_PLAYER_ATTRIBUTE_ALTERATION_STATUSES,
};