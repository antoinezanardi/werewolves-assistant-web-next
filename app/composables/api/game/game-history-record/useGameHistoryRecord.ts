import type { GameHistoryRecord } from "~/composables/api/game/types/game-history-record/game-history-record.class";

type UseGameHistoryRecord = {
  didSourceSkipped: (gameHistoryRecord: GameHistoryRecord) => boolean;
};

function useGameHistoryRecord(): UseGameHistoryRecord {
  function didSourceSkipped(gameHistoryRecord: GameHistoryRecord): boolean {
    const {
      type,
      votes,
      targets,
      didJudgeRequestAnotherVote,
      chosenCard,
    } = gameHistoryRecord.play;

    return type === "vote" && !votes ||
      type === "target" && !targets ||
      type === "request-another-vote" && didJudgeRequestAnotherVote === undefined ||
      type === "choose-card" && !chosenCard;
  }
  return { didSourceSkipped };
}

export { useGameHistoryRecord };