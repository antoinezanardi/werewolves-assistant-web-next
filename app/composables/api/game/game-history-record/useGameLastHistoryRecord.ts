import type { MaybeRef, Ref } from "vue";
import type { GameHistoryRecordPlayerAttributeAlterationStatus } from "~/composables/api/game/types/game-history-record/game-history-record-player-attribute-alteration/game-history-record-player-attribute-alteration.types";
import type { GameHistoryRecord } from "~/composables/api/game/types/game-history-record/game-history-record.class";
import type { Game } from "~/composables/api/game/types/game.class";
import type { GameSource } from "~/composables/api/game/types/game.types";
import type { PlayerAttributeName } from "~/composables/api/game/types/players/player-attribute/player-attribute.types";
import type { Player } from "~/composables/api/game/types/players/player.class";

type UseGameLastHistoryRecord = {
  lastTargetedPlayers: Ref<Player[]>;
  doesHavePlayerAttributeAlteration: (attributeName: PlayerAttributeName, source: GameSource, status: GameHistoryRecordPlayerAttributeAlterationStatus) => boolean;
};

function useGameLastHistoryRecord(game: MaybeRef<Game>): UseGameLastHistoryRecord {
  const lastGameHistoryRecord = computed<GameHistoryRecord | null>(() => (isRef(game) ? game.value.lastGameHistoryRecord : game.lastGameHistoryRecord));

  const lastTargetedPlayers = computed<Player[]>(() => {
    if (!lastGameHistoryRecord.value?.play.targets) {
      return [];
    }
    return lastGameHistoryRecord.value.play.targets.map(target => target.player);
  });

  function doesHavePlayerAttributeAlteration(attributeName: PlayerAttributeName, source: GameSource, status: GameHistoryRecordPlayerAttributeAlterationStatus): boolean {
    if (!lastGameHistoryRecord.value?.playerAttributeAlterations) {
      return false;
    }
    return lastGameHistoryRecord.value.playerAttributeAlterations.some(playerAttributeAlteration => playerAttributeAlteration.name === attributeName &&
      playerAttributeAlteration.source === source &&
      playerAttributeAlteration.status === status);
  }
  return {
    lastTargetedPlayers,
    doesHavePlayerAttributeAlteration,
  };
}

export { useGameLastHistoryRecord };