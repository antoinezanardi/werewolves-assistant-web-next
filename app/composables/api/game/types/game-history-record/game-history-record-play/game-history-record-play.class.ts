import { Expose, plainToInstance, Type } from "class-transformer";

import { GameAdditionalCard } from "~/composables/api/game/types/game-additional-card/game-additional-card.class";
import { GameHistoryRecordPlaySource } from "~/composables/api/game/types/game-history-record/game-history-record-play/game-history-record-play-source/game-history-record-play-source.class";
import { GameHistoryRecordPlayTarget } from "~/composables/api/game/types/game-history-record/game-history-record-play/game-history-record-play-target/game-history-record-play-target.class";
import { GameHistoryRecordPlayVote } from "~/composables/api/game/types/game-history-record/game-history-record-play/game-history-record-play-vote/game-history-record-play-vote.class";
import { GameHistoryRecordPlayVoting } from "~/composables/api/game/types/game-history-record/game-history-record-play/game-history-record-play-voting/game-history-record-play-voting.class";
import type { GamePlayAction, GamePlayCause, GamePlayType } from "~/composables/api/game/types/game-play/game-play.types";
import type { RoleSide } from "~/composables/api/role/types/role.types";
import { DEFAULT_PLAIN_TO_INSTANCE_OPTIONS } from "~/utils/constants/class-transformer.constants";

class GameHistoryRecordPlay {
  @Expose()
  public type: GamePlayType;

  @Expose()
  public action: GamePlayAction;

  @Expose()
  @Type(() => GameHistoryRecordPlaySource)
  public source: GameHistoryRecordPlaySource;

  @Expose()
  public cause?: GamePlayCause;

  @Expose()
  @Type(() => GameHistoryRecordPlayTarget)
  public targets?: GameHistoryRecordPlayTarget[];

  @Expose()
  @Type(() => GameHistoryRecordPlayVote)
  public votes?: GameHistoryRecordPlayVote[];

  @Expose()
  @Type(() => GameHistoryRecordPlayVoting)
  public voting?: GameHistoryRecordPlayVoting;

  @Expose()
  public didJudgeRequestAnotherVote?: boolean;

  @Expose()
  @Type(() => GameAdditionalCard)
  public chosenCard?: GameAdditionalCard;

  @Expose()
  public chosenSide?: RoleSide;

  public static create(gameHistoryRecordPlay: GameHistoryRecordPlay): GameHistoryRecordPlay {
    return plainToInstance(GameHistoryRecordPlay, gameHistoryRecordPlay, DEFAULT_PLAIN_TO_INSTANCE_OPTIONS);
  }
}

export { GameHistoryRecordPlay };