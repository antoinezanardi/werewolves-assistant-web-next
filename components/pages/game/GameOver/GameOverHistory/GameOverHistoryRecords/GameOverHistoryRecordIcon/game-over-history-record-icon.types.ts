import type { GameHistoryRecord } from "~/composables/api/game/types/game-history-record/game-history-record.class";

type GameOverHistoryRecordIconProps = {
  gameHistoryRecord: GameHistoryRecord;
};

export type { GameOverHistoryRecordIconProps };