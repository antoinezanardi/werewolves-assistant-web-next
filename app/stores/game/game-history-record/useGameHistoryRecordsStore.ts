import type { AsyncDataRequestStatus } from "nuxt/app";
import { defineStore } from "pinia";

import { useFetchGameHistoryRecords } from "~/composables/api/game/game-history-record/useFetchGameHistoryRecords";
import type { GameHistoryRecord } from "~/composables/api/game/types/game-history-record/game-history-record.class";
import { StoreIds } from "~/stores/enums/store.enum";

const useGameHistoryRecordsStore = defineStore(StoreIds.GAME_HISTORY_RECORDS, () => {
  const { getGameHistoryRecords } = useFetchGameHistoryRecords();

  const gameHistoryRecords = ref<GameHistoryRecord[]>([]);
  const fetchingGameHistoryRecordsStatus = ref<AsyncDataRequestStatus>("idle");

  async function fetchAndSetGameHistoryRecords(gameId: string): Promise<void> {
    fetchingGameHistoryRecordsStatus.value = "pending";
    const fetchedGameHistoryRecords = await getGameHistoryRecords(gameId);
    if (!fetchedGameHistoryRecords) {
      fetchingGameHistoryRecordsStatus.value = "error";

      return;
    }
    gameHistoryRecords.value = fetchedGameHistoryRecords;
    fetchingGameHistoryRecordsStatus.value = "success";
  }
  return {
    gameHistoryRecords,
    fetchingGameHistoryRecordsStatus,
    fetchAndSetGameHistoryRecords,
  };
});

export { useGameHistoryRecordsStore };