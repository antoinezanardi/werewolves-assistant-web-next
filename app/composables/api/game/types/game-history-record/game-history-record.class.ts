import { Expose, plainToInstance, Type } from "class-transformer";

import { GameHistoryRecordPlay } from "~/composables/api/game/types/game-history-record/game-history-record-play/game-history-record-play.class";
import { GameHistoryRecordPlayerAttributeAlteration } from "~/composables/api/game/types/game-history-record/game-history-record-player-attribute-alteration/game-history-record-player-attribute-alteration.class";
import { GamePhase } from "~/composables/api/game/types/game-phase/game-phase.class";
import { Player } from "~/composables/api/game/types/players/player.class";
import { DEFAULT_PLAIN_TO_INSTANCE_OPTIONS } from "~/utils/constants/class-transformer.constants";

class GameHistoryRecord {
  @Expose()
  public _id: string;

  @Expose()
  public gameId: string;

  @Expose()
  public turn: number;

  @Type(() => GamePhase)
  @Expose()
  public phase: GamePhase;

  @Expose()
  public tick: number;

  @Expose()
  @Type(() => GameHistoryRecordPlay)
  public play: GameHistoryRecordPlay;

  @Expose()
  @Type(() => Player)
  public revealedPlayers?: Player[];

  @Expose()
  @Type(() => Player)
  public deadPlayers?: Player[];

  @Expose()
  @Type(() => GameHistoryRecordPlayerAttributeAlteration)
  public playerAttributeAlterations?: GameHistoryRecordPlayerAttributeAlteration[];

  @Expose()
  public createdAt: Date;

  public static create(gameHistoryRecord: GameHistoryRecord): GameHistoryRecord {
    return plainToInstance(GameHistoryRecord, gameHistoryRecord, DEFAULT_PLAIN_TO_INSTANCE_OPTIONS);
  }
}

export { GameHistoryRecord };