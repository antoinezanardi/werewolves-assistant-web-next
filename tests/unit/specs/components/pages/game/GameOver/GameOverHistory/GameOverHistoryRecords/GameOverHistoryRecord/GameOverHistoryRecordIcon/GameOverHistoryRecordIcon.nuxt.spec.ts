import type { mount } from "@vue/test-utils";
import type { ComponentMountingOptions } from "@vue/test-utils/dist/mount";

import type { GameOverHistoryRecordIconProps } from "~/components/pages/game/GameOver/GameOverHistory/GameOverHistoryRecords/GameOverHistoryRecordIcon/game-over-history-record-icon.types";
import GameOverHistoryRecordIcon from "~/components/pages/game/GameOver/GameOverHistory/GameOverHistoryRecords/GameOverHistoryRecordIcon/GameOverHistoryRecordIcon.vue";
import { createFakeGameHistoryRecordPlaySource } from "~/tests/unit/utils/factories/composables/api/game/game-history-record/game-history-record-play/game-history-record-play-source/game-history-record-play-source.factory";
import { createFakeGameHistoryRecordPlay } from "~/tests/unit/utils/factories/composables/api/game/game-history-record/game-history-record-play/game-history-record-play.factory";
import { createFakeGameHistoryRecord } from "~/tests/unit/utils/factories/composables/api/game/game-history-record/game-history-record.factory";
import { mountSuspendedComponent } from "~/tests/unit/utils/helpers/mount.helpers";

describe("Game Over History Record Icon Component", () => {
  const gameHistoryRecord = createFakeGameHistoryRecord({
    play: createFakeGameHistoryRecordPlay({
      source: createFakeGameHistoryRecordPlaySource({ name: "sheriff" }),
      action: "delegate",
    }),
  });
  let wrapper: ReturnType<typeof mount<typeof GameOverHistoryRecordIcon>>;
  const defaultProps: GameOverHistoryRecordIconProps = { gameHistoryRecord };

  async function mountGameOverHistoryRecordIconComponent(options: ComponentMountingOptions<typeof GameOverHistoryRecordIcon> = {}):
  Promise<ReturnType<typeof mount<typeof GameOverHistoryRecordIcon>>> {
    return mountSuspendedComponent(GameOverHistoryRecordIcon, {
      props: defaultProps,
      ...options,
    });
  }

  beforeEach(async() => {
    wrapper = await mountGameOverHistoryRecordIconComponent();
  });

  it("should match snapshot when rendered.", () => {
    expect(wrapper).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe("Icon", () => {
    it("should display the game history record icon when rendered.", () => {
      const icon = wrapper.find<HTMLImageElement>("[alt='Game history record icon']");

      expect(icon.attributes("src")).toBe("/svg/game/player/player-attribute/sheriff.svg");
    });
  });
});