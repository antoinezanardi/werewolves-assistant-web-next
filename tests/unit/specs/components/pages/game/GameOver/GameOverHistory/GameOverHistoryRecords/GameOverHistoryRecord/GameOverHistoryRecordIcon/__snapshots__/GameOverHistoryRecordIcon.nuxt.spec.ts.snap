// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Game Over History Record Icon Component > should match snapshot when rendered. 1`] = `
"<div id="game-history-record-icon" class="align-items-center flex h-12 justify-content-center shadow-1 text-white w-12 z-1">
  <nuxt-img-stub alt="Game history record icon" src="/svg/game/player/player-attribute/sheriff.svg" preload="false" crossorigin="false" placeholder="/svg/misc/ripples.svg"></nuxt-img-stub>
</div>"
`;
